driver-sync/src/main/com/mongodb/client/internal/ClientSessionBinding.java

```diff
@@ -32,21 +38,46 @@
  * This class is not part of the public API and may be removed or changed at any time.
  */
 public class ClientSessionBinding implements ReadWriteBinding {
-    private final ReadWriteBinding wrapped;
+    private ReadWriteBinding wrapped;
     private final ClientSession session;
     private final boolean ownsSession;
     private final ClientSessionContext sessionContext;
+    private final Cluster cluster;
+    private ReadPreference readPreference;
 
     public ClientSessionBinding(final ClientSession session, final boolean ownsSession, final ReadWriteBinding wrapped) {
-        this.wrapped = notNull("wrapped", wrapped);
-        this.ownsSession = ownsSession;
         this.session = notNull("session", session);
+        this.ownsSession = ownsSession;
         this.sessionContext = new SyncClientSessionContext(session);
+        this.cluster = wrapped instanceof ClusterBinding ? ((ClusterBinding) wrapped).getCluster() : null;
+        this.wrapped = notNull("wrapped", initShardedTxnWrapped(wrapped));
+    }
+
+    private ReadWriteBinding initShardedTxnWrapped(final ReadWriteBinding wrapped) {
+        if (isActiveShardedTxn()) {
+            this.readPreference = session.getTransactionOptions().getReadPreference();
```

Comment:
This should be the same as ClusterBinding's read preference.  You can simplify the intent by just grabbing it from there and removing the extra tracking field.

Reply:
One more though there.  This whole method seems redundant with `setWrappedOnPinnedMongosReset`.  Does everything just work if you get rid of this usage in the constructor and just delay until you need a connection source?

Reply:
Done

Reply:
I tried delaying pinning until getting the connection source, but that produced errors in the tests. I have removed `initShardedTxnWrapped` and now initialize in the constructor by calling `setWrappedOnPinnedMongosReset`.

Reply:
Any hypothesis on why removing from the constructor causes a failure?  I can't think of a reason why that would be.

Reply:
I believe I have resolved the issue. I was not releasing the allocated bindings after getting the connection source. Once I was able to restructure the code to release the allocated bindings, the transaction test errors resolved.